### PR TITLE
feat(tools): add Terraform fmt formatter and validate linter

### LINKED ISSUE #1515: feat(tools): add Terraform support (terraform fmt + validate)
## Summary

Add first-class Terraform support to lintro: `terraform fmt` as a fixable formatter and `terraform validate` as a check-only linter. This lets repos with `*.tf` code (e.g. Rustume's `infra/` tree, moving to Rustume-ops) run their IaC checks through `lintro chk` / `lintro fmt` instead of hand-rolled CI scripts.

## Problem

Terraform is absent from lintro's tool matrix. Downstream repos currently invoke the native binary directly in CI (e.g. Rustume's `scripts/ci/terraform/validate.sh` running `terraform fmt -check` + `terraform validate` per module/root), which bypasses lintro's unified reporting, config, and local/CI parity.

## Proposed Solution

Two capabilities behind one tool plugin (or a formatter + linter pair, whichever fits the plugin model better):

- **fmt**: `terraform fmt -check -recursive` for check mode; `terraform fmt -recursive` for `lintro fmt` (auto-fix ✅). Parse the changed/offending file list from stdout.
- **validate**: per Terraform root/module, run `terraform init -backend=false -input=false` then `terraform validate -json` and map the JSON diagnostics (severity, range, summary/detail) to lintro issues.

Discovery: treat any directory containing `.tf` files as a target for fmt; for validate, treat directories with `.tf` files as roots/modules (dedupe nested `.terraform/`).

## Implementation Notes

- Follow the standard new-tool plugin flow (`lintro/tools/implementations/`, definitions, manifest entry, `lintro doctor` / `lintro install` wiring).
- Install source: HashiCorp releases / `brew install terraform`; consider honoring an already-present `tofu` binary later (out of scope here).
- `validate` needs network-free init (`-backend=false`); provider plugins may still need downloading — respect a warm `.terraform.lock.hcl` and document the caching story for CI.
- `-json` output for validate makes parsing robust; `fmt` has no JSON mode, parse the file list.
- Related: #422 (checkov) covers IaC *security* scanning — separate concern, keep both.

## Benefits

- IaC formatting and validity checks in the same `lintro chk`/`fmt` invocation as everything else.
- Removes bespoke terraform CI scripts downstream.

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Commit Summary (Conventional Commits)

- Title (required, present tense):

  ```text
  feat(tools): add Terraform fmt formatter and validate linter
  ```

- Type:
  - [x] feat (minor)
  - [ ] fix / perf (patch)
  - [ ] docs
  - [ ] refactor
  - [ ] test
  - [ ] chore / ci / style

- Breaking change:
  - [ ] `!` in title or `BREAKING CHANGE:` footer included

### Release Trigger Rules (exact)

- A merged PR will bump the version based on its title (squash merge required):
  - `feat(...)` or `feat:` → MINOR bump
  - `fix(...)` / `fix:` or `perf(...)` / `perf:` → PATCH bump
  - Any title with `!` after the type (e.g. `feat!:` or `feat(scope)!:`) or a body
    containing `BREAKING CHANGE:` → MAJOR bump
- Use squash merge so the PR title becomes the merge commit title.

## What’s Changing

Adds first-class Terraform support:

- `terraform fmt` — check/fix via lintro (parse offending file list)
- `terraform validate` — per-module `init -backend=false` then `validate -json`, mapped to lintro issues
- Full lintro-add wiring: plugin, parser, manifest, install-tools, Docker verify, doctor, tests, samples, docs, site

CI follow-ups on this PR: HashiCorp egress allowlist, prettier 3.9.4, drop Rust HTML docs from tools image to avoid Trivy scan timeout.

## Checklist

- [x] Title follows Conventional Commits
- [x] Tests added/updated
- [x] Docs updated if user-facing
- [ ] Local CI passed (`./scripts/local/run-tests.sh`)

## Closes

- Closes #1515

## Details

_See What’s Changing._

## Summary by CodeRabbit

* **New Features**
  * Added Terraform as a supported linting and formatting tool.
  * Added Terraform formatting checks and automatic fixes.
  * Added optional Terraform validation with diagnostics for configuration issues.
  * Added Terraform installation support, version detection, and documentation.
* **Documentation**
  * Added configuration, usage, analysis, and setup guidance for Terraform.
* **Bug Fixes**
  * Improved Docker workflow connectivity and pinned build tooling for reliable builds.

<details><summary><h3>Greptile Summary</h3></summary>

Adds first-class Terraform support to lintro by wrapping `terraform fmt` (check and in-place fix) and `terraform validate` (per-module `init -backend=false` followed by `-json` diagnostics). The integration touches the full plugin lifecycle: parser, issue model, tool definition, manifest, install script, Renovate tracking, Docker verification, and a comprehensive unit + integration test suite.

- **`TerraformPlugin`** orchestrates `terraform fmt -check` for formatting issues and, optionally, per-module `terraform init` + `terraform validate -json` for configuration diagnostics; timeout, init-failure, and parse-error paths are all explicitly handled.
- **`terraform_parser.py`** cleanly separates the line-based fmt parser from the JSON validate parser, and the root-module path fallback (`normalized_dir or module_dir or "."`) correctly handles both `""` and `"."` inputs — the concern raised in a previous review comment is already resolved.
- **Install script** now includes the version-compare-and-upgrade pattern (following `shellcheck`/`osv-scanner`), addressing the thread from the earlier review.
</details>

<h3>Confidence Score: 5/5</h3>

- Safe to merge. The new Terraform plugin is well-isolated, follows the existing plugin contract, and all edge cases (timeout, init failure, parse failure) are explicitly handled.
- The change adds a self-contained new tool plugin with no modifications to shared core logic beyond enum and registry registration. Tests cover both the happy path and all documented error paths using mocked subprocesses. The one fragility noted — malformed validate stdout not surfacing a fallback error — is an extremely rare crash scenario that `terraform validate -json` virtually never triggers in practice.
- No files require special attention. The `_run_validate` error-fallback logic in `terraform.py` could be hardened, but this is a very low-risk edge case.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| lintro/tools/definitions/terraform.py | New TerraformPlugin covering `terraform fmt` (check/fix) and `terraform validate` (per-module init + JSON diagnostics). Logic is well-structured, timeout/init-failure/parse-error paths are all handled, and the fixed-count arithmetic is correct. One edge-case fragility: a non-zero validate exit with non-empty but non-JSON stdout is silently dropped. |
| lintro/parsers/terraform/terraform_parser.py | Two-format parser (line-based for fmt, JSON for validate). Handles None/empty/whitespace/invalid-JSON gracefully. Root-module path fallback (`file_path = normalized_dir or module_dir or "."`) correctly covers both `""` and `"."` inputs. Previous concerns about empty file fields are resolved. |
| scripts/utils/install-tools.sh | Terraform install block follows the same checksum-pinned binary-install pattern as other tools. Now includes the version comparison + upgrade path requested in the previous review thread. Inner function is properly cleaned up on all exit paths with `rm -rf $tmpdir`. |
| tests/unit/tools/terraform/test_execution.py | Comprehensive mocked-subprocess unit tests for check/fix flows, including timeout, init failure, and validate-diagnostic paths. Uses SubprocessResult correctly for the validate path and (True, ...) tuples for the init path, mirroring production code. |
| tests/unit/parsers/test_terraform_parser.py | Good coverage of both parsers including edge cases (None, whitespace, invalid JSON, root module `""` vs `"."`, warning without range, base_dir prefix). Parametrized tests cover all relevant code branches. |
| tests/integration/tools/terraform/conftest.py | Integration fixtures skip cleanly when the binary is absent. Copies sample files into isolated tmp directories so integration runs don't mutate committed fixtures. |
| lintro/parsers/terraform/terraform_issue.py | Minimal dataclass extending BaseIssue with `level` and `code` fields. DISPLAY_FIELD_MAP correctly remaps `level` to the `"severity"` display key. |
| renovate.json | Renovate regex rule added to track `ToolName.TERRAFORM` version in `_tool_versions.py` against the `hashicorp/terraform` GitHub releases datasource. Pattern and extractVersionTemplate look correct. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[check / fix called] --> B[_prepare_execution\nresolve files, cwd, timeout]
    B --> C{rel_files\nempty?}
    C -- yes --> D[return early ToolResult]
    C -- no --> E[_run_fmt_check\nterraform fmt -check *rel_files]
    E --> F{TimeoutExpired?}
    F -- yes --> G[timeout TerraformIssue]
    F -- no --> H[parse_terraform_fmt_output\nlines ending .tf/.tfvars]
    H --> I{validate\noption?}
    I -- false --> J[skip validate]
    I -- true --> K[_module_dirs\nunique dirs from .tf files\nexclude .terraform cache]
    K --> L[for each module_dir\nterraform init -backend=false]
    L --> M{init ok?}
    M -- no --> N[init error TerraformIssue]
    M -- yes --> O[terraform validate -json]
    O --> P{TimeoutExpired?}
    P -- yes --> Q[timeout TerraformIssue]
    P -- no --> R[parse_terraform_validate_output\nJSON diagnostics → issues]
    R --> S{parsed\nempty + failed?}
    S -- stdout empty --> T[fallback error TerraformIssue]
    S -- no --> U[collect issues]
    J --> V[assemble ToolResult]
    G --> V
    N --> V
    Q --> V
    T --> V
    U --> V
    V --> W{fix mode?}
    W -- yes --> X[terraform fmt *rel_files\napply in-place]
    X --> Y[_run_fmt_check again\nremaining issues]
    Y --> Z[ToolResult with\ninitial / fixed / remaining counts]
    W -- no --> AA[ToolResult\ncheck result]
```
</details>

<sub>Reviews (5): Last reviewed commit: ["fix(terraform): repair merge scars, fix ..."](https://github.com/lgtm-hq/py-lintro/commit/5ec239d903743590e760b6b8767c299a88cc20c3) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=45363969)</sub>

### CODE REVIEW

File: scripts/utils/install-tools.sh [hunk 1]
```diff
@@ -1533,6 +1534,79 @@ main() {
...
+	if should_install "terraform"; then
+		# Install terraform (infrastructure-as-code formatter and validator)
+		# Prebuilt binaries: https://releases.hashicorp.com/terraform
+		echo -e "${BLUE}Installing terraform...${NC}"
+		TERRAFORM_VERSION=$(get_tool_version "terraform") || exit 1
+		if [ $DRY_RUN -eq 1 ]; then
+			log_info "[DRY-RUN] Would install terraform v${TERRAFORM_VERSION}"
+		elif command -v terraform &>/dev/null; then
+			echo -e "${GREEN}✓ terraform already installed${NC}"
+		else
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/utils/install-tools.sh]: Fixed in 010d9981. The terraform block now mirrors the shellcheck/osv-scanner check-and-upgrade pattern: when terraform is already on PATH we parse its version and compare with version_ge against the pinned TERRAFORM_VERSION (1.15.8). If it is older (or unparseable) we install the pinned build; if current-or-newer we skip. Verified end-to-end locally: a pre-existing terraform v1.5.7 was detected as older and upgraded to v1.15.8 with checksum verification.

File: scripts/utils/install-tools.sh [hunk 2]
```diff
@@ -1533,6 +1534,79 @@ main() {
...
+			checksum_url="${base_url}/terraform_${TERRAFORM_VERSION}_SHA256SUMS"
+			if download_with_retries "$zip_url" "$tmpdir/terraform.zip" 3; then
+				# Require checksum verification before installing
+				if ! download_with_retries "$checksum_url" "$tmpdir/checksums.txt" 3; then
+					echo -e "${RED}✗ Failed to download checksum file for terraform${NC}"
+					rm -rf "$tmpdir"
+					exit 1
+				fi
+				echo -e "${BLUE}Verifying checksum for terraform...${NC}"
+				expected=$(grep "terraform_${TERRAFORM_VERSION}_${os}_${arch_name}.zip" "$tmpdir/checksums.txt" | awk '{print $1}')
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/utils/install-tools.sh:1750]: Not changing this here. Checksum-pinned verification against the release publishers SHA256SUMS is the uniform convention every binary installer in this script uses (osv-scanner, shellcheck, taplo, etc.); terraform follows the same shape. Adding a GPG trust chain — pinning HashiCorps signing key and requiring gpg in every install environment — is a script-wide security enhancement that should apply to all installers uniformly, so it belongs in its own dedicated issue rather than being introduced piecemeal in this terraform PR. Note the expected checksum here is not repo-pinned; it is taken from the SHA256SUMS file downloaded alongside the archive, which is exactly why standardizing signature verification is worth doing consistently across all installers rather than one-off.

File: lintro/tools/definitions/terraform.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+            "fmt",
+            *ctx.rel_files,
+        ]
+        logger.debug(f"[TerraformPlugin] fix: {' '.join(fix_cmd)} (cwd={ctx.cwd})")
+        try:
+            self._run_subprocess(cmd=fix_cmd, timeout=ctx.timeout, cwd=ctx.cwd)
+        except subprocess.TimeoutExpired:
+            timeout_issue = self._timeout_issue("terraform fmt", ctx.timeout)
+            initial_issues = initial_fmt_issues + validate_issues
+            remaining = initial_issues + [timeout_issue]
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/tools/definitions/terraform.py]: Valid — fixed in 4b21e462. The TimeoutExpired branch of fix() now builds initial_issues = initial_fmt_issues + validate_issues + [timeout_issue] and sets remaining = list(initial_issues), so initial == fixed + remaining holds with fixed_issues_count = 0 and issues_count == remaining_issues_count. Covered by the new test_fix_fmt_apply_timeout_keeps_counts_consistent in tests/unit/tools/terraform/test_execution.py.

File: lintro/parsers/terraform/terraform_parser.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            file_path = normalized_dir
+
+        issues.append(
+            TerraformIssue(
+                file=file_path,
+                line=line,
+                column=column,
+                level=severity,
+                code="validate",
+                message=message,
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/parsers/terraform/terraform_parser.py]: Valid — fixed in 4b21e462. parse_terraform_validate_output now uses file_path = normalized_dir or module_dir or '.' in the no-range branch, so a root-module diagnostic reports '.' instead of an empty file field. Covered by the new parametrized test_parse_validate_output_root_module_without_range (module_dir '' and '.') in tests/unit/parsers/test_terraform_parser.py.

File: tests/unit/tools/terraform/test_execution.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    other.write_text("not terraform")
+
+    with patch(
+        "lintro.plugins.execution_preparation.verify_tool_version",
+        return_value=None,
+    ):
+        result = terraform_plugin.fix([str(other)], {})
+
+    assert_that(result.success).is_true()
+    assert_that(result.output).contains("No")
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply tests/unit/tools/terraform/test_execution.py]: Valid — added in 4b21e462. test_fix_fmt_apply_timeout_keeps_counts_consistent patches _run_subprocess with side_effect=[(False, 'main.tf'), subprocess.TimeoutExpired(...)] and asserts fixed_issues_count == 0, initial_issues_count == remaining_issues_count, issues_count == remaining_issues_count, and that a 'timeout' code is present. It fails against the pre-fix accounting.

File: tests/unit/tools/terraform/test_options.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    value: object,
+) -> None:
+    """Non-boolean validate values raise ValueError.
+
+    Args:
+        terraform_plugin: The TerraformPlugin instance to test.
+        value: An invalid value for validate.
+    """
+    with pytest.raises(ValueError, match="validate must be a boolean"):
+        terraform_plugin.set_options(validate=value)  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply tests/unit/tools/terraform/test_options.py]: Partly valid — addressed in 4b21e462 by adding public-surface coverage rather than deleting the unit test. New test_check_inits_each_module_directory_once in tests/unit/tools/terraform/test_execution.py drives check() over a/main.tf, b/main.tf and b/.terraform/cached.tf, then asserts the init invocations (read off the _run_subprocess call args) are exactly ['a', 'b'] — so cache-skipping and per-module dedup are now verified through check(). The parametrized _module_dirs test is kept as a cheap table-driven check of the same logic.

**Comment by TurboCoder13** [comment]
Parked (2026-09-04, owner decision) — converted to draft and labelled `blocked`.

The tool-definition layer is being factored (#2311 in the re-alignment roadmap #2288): shared check/fix scaffolding is lifted into `lintro/tools/core`, definitions move to one package per tool, and a duplicate-code gate (#2293) now fails any PR that adds another copy of the current template. This PR was written against the current template and would add ~300 duplicated lines plus ~30 central edits.

When #2311's acceptance criteria hold, a comment will land here with the new shape; the executing agent then rebases this PR if the conflict set is smaller than the diff, or closes it and re-implements from its issue. No action needed until then.

**Comment by TurboCoder13** [comment]
Phase 2 of #2288 (un-park after C.3 #2311): the package-per-tool layout is now on `main` (e0f3786c). A tool lives in `lintro/tools/<tool>/{__init__.py, definition.py}` (no `lintro/tools/definitions/` anymore); discovery scans per-tool packages; `lintro/plugins/_builtin_index.py` is regenerated with `python3 scripts/ci/generate-builtin-tool-index.py`; the package must be appended to `[tool.lintro.pylint] include` and `GATE_PACKAGES` in tests/unit/test_duplicate_code_baseline.py (duplicate-code gate is zero-tolerance); every definition declares `claims`/`reads_tree`/`partitionable` (#1740, PR #2420, landing next); shared per-file/batch runners in `lintro/tools/core/{check_runner,fix_runner,batch_runner}.py` replace hand-rolled loops. See docs/contributing/adding-a-new-tool.md. This PR will be rebased onto that layout one at a time, or closed and re-implemented from its issue if the rebase is larger than the original diff.
